app/components-react/editor/layouts/hooks.tsx

```diff
@@ -0,0 +1,138 @@
+    window.addEventListener('resize', () => updateSize());
+
+    return () => {
+      window.removeEventListener('resize', () => updateSize());
+    };
+  }, [chatCollapsed]);
+
+  const componentEl = useRef<HTMLDivElement | null>(null);
+
+  const componentRef = useCallback(node => {
+    if (node) {
+      componentEl.current = node;
+      onTotalWidth(mapVectors(vectors), isColumns);
+      updateSize();
+    }
+  }, []);
+
+  function vectorsToSlots() {
+    const slotArray: Array<ILayoutSlotArray> = [];
+    vectors.forEach(vector => {
+      if (typeof vector === 'string') slotArray.push([vector]);
+      else if (Array.isArray(vector)) slotArray.push(vector);
+    });
+    return slotArray;
+  }
+
+  function getBarPixels(bar: 'bar1' | 'bar2', offset: number) {
+    if (!componentEl.current) return;
+    // Migrate from pixels to proportions
+    if ((resizes[bar] as number) >= 1) setBar(bar, resizes[bar] as number);
+    const { height, width } = componentEl.current.getBoundingClientRect();
+    const offsetSize = isColumns ? width - offset : height;
+    return Math.round(offsetSize * (resizes[bar] as number));
+  }
+
+  function setBar(bar: 'bar1' | 'bar2', val: number) {
+    if (val === 0 || !componentEl.current) return;
+    setBars(oldState => ({ ...oldState, [bar]: val }));
+    const { height, width } = componentEl.current.getBoundingClientRect();
+    const totalSize = isColumns ? width : height;
+    const proportion = parseFloat((val / totalSize).toFixed(2));
+    LayoutService.actions.setBarResize(bar, proportion);
+  }
+
+  function minsFromSlot(slot: LayoutSlot) {
+    // If there is no component slotted we return no minimum
+    if (!childrenMins || !childrenMins[slot]) return { x: 0, y: 0 };
+    return childrenMins[slot];
+  }
+
+  function calculateMinimum(slots: ILayoutSlotArray) {
+    const mins = mapVectors(slots);
+    return calculateMin(mins);
+  }
+
+  function mapVectors(slots: ILayoutSlotArray): IVec2Array {
+    return slots.map(slot => {
+      if (Array.isArray(slot)) return mapVectors(slot);
+      return minsFromSlot(slot);
+    });
+  }
+
+  function calculateMin(slots: IVec2Array) {
+    return LayoutService.views.calculateMinimum(isColumns ? 'x' : 'y', slots);
+  }
+
+  function updateSize(chatCollapsed = true, oldChatCollapsed?: boolean) {
+    let offset = chatCollapsed ? 0 : livedockSize;
+    // Reverse offset if chat is collapsed from an uncollapsed state
+    if (chatCollapsed && oldChatCollapsed === false) {
+      offset = livedockSize * -1;
+    }
+    const bar1 = getBarPixels('bar1', offset);
+    if (bar1) setBar('bar1', bar1);
+    if (mins.bar2) {
+      const bar2 = getBarPixels('bar2', offset);
+      if (bar2) setBar('bar2', bar2);
+    }
+  }
+
+  function calculateMax(restMin: number) {
+    if (!componentEl.current) return 0;
+    const { height, width } = componentEl.current.getBoundingClientRect();
+    const max = isColumns ? width : height;
+    return max - restMin;
+  }
+
```

Comment:
I suggest using `useCallback` for each function inside components that are always visible in our app. Otherwise, this entire list of functions will be recreated for each `useLayout` call. That leads to performance issues.